roles/jd-server/src/lib/mempool/mod.rs

```diff
@@ -58,14 +59,34 @@ impl JDsMempool {
         }
     }
 
-    pub fn add_tx_data_to_mempool(
+    // this functions fill in the mempool the transactions with the given txid and insert the given
+    // transactions. The ids are for the transactions that are already known to the node, the
+    // unknown transactions are provided directly as a vector
+    pub async fn add_tx_data_to_mempool(
         self_: Arc<Mutex<Self>>,
-        txid: Txid,
-        transaction: Option<Transaction>,
-    ) {
-        let _ = self_.safe_lock(|x| {
-            x.mempool.insert(txid, transaction);
-        });
+        add_txs_to_mempool_inner: AddTrasactionsToMempoolInner,
+    ) -> Result<(), JdsMempoolError> {
+        let txids = add_txs_to_mempool_inner.known_transactions;
+        let transactions = add_txs_to_mempool_inner.unknown_transactions;
+        let client = self_
+            .safe_lock(|a| a.get_client())
+            .map_err(|e| JdsMempoolError::PoisonLock(e.to_string()))?
+            .ok_or(JdsMempoolError::NoClient)?;
+        // fill in the mempool the transactions id in the mempool with the full transactions
+        // retrieved from the jd client
+        for txid in txids {
+            let transaction = client
+                .get_raw_transaction(&txid.to_string(), None)
```

Comment:
before fetching txs you should check if they are already in the mempool

Reply:
another option could  be to use an HashSet right after the receiver

Reply:
I am struggling to figure out a useful case for HashSet. If you propose to use it instead HashMap, then you cant store transactions only by their IDs. What do you mean with "right after the receiver"? Can you be a bit more specific? ^^

BTW already fixed in [this commit](https://github.com/lorbax/stratum/commit/d99d438268660bf31bc4d10cd7c30f7244887141)